contracts/SwapPaymentTerminal.sol

```diff
@@ -0,0 +1,310 @@
+
+  /// @notice Logic to be triggered if a transfer should be undone
+  /// @param _to The address to which the transfer went.
+  /// @param _amount The amount of the transfer, as a fixed point number with the same number of decimals as this terminal.
+  function _cancelTransferTo(address _to, uint256 _amount) internal override {
+    IERC20(token).safeDecreaseAllowance(_to, _amount);
+  }
+
+  /// @notice Contribute any tokens to a project.
+  /// @param _projectId The ID of the project being paid.
+  /// @param _amount The amount of terminal tokens being received, as a fixed point number with the same amount of decimals as this terminal. If this terminal's token is ETH, this is ignored and msg.value is used in its place.
+  /// @param _token The token being paid, that'll be converted to the token the project wants via an AMM swap if they don't already match.
+  /// @param _beneficiary The address to mint tokens for and pass along to the funding cycle's data source and delegate.
+  /// @param _minReturnedTokens The minimum number of project tokens expected in return, as a fixed point number with the same amount of decimals as this terminal.
+  /// @param _preferClaimedTokens A flag indicating whether the request prefers to mint project tokens into the beneficiaries wallet rather than leaving them unclaimed. This is only possible if the project has an attached token contract. Leaving them unclaimed saves gas.
+  /// @param _memo A memo to pass along to the emitted event, and passed along the the funding cycle's data source and delegate.  A data source can alter the memo before emitting in the event and forwarding to the delegate.
+  /// @param _metadata Bytes to send along to the data source, delegate, and emitted event, if provided.
+  /// @return The number of tokens minted for the beneficiary, as a fixed point number with 18 decimals.
+  function pay(
+    uint256 _projectId,
+    uint256 _amount,
+    address _token,
+    address _beneficiary,
+    uint256 _minReturnedTokens,
+    bool _preferClaimedTokens,
+    string calldata _memo,
+    bytes calldata _metadata
+  ) public payable virtual override returns (uint256) {
+    // The goal is to convert the incoming token into the one we can accept.
+    // if the incoming token is the one we accept, do nothing.
+    // else, try to swap the incoming token into the one we accept.
+
+    // If the incoming token isn't the one we want, swap.
+    if (_token != token) {
+      address _payer;
+      // Wrap ETH into WETH if relevant.
+      if (_token == JBTokens.ETH) {
+        WETH.deposit{value: _amount}();
+        _payer = address(this);
+      } else {
+        _payer = msg.sender;
+      }
+
+      uint256 _fee;
+
+      // Unpack the quote from the pool, given by the frontend.
+      (bool _quoteExists, _metadata) = JBDelegateMetadataLib.getMetadata(123, _metadata);
+      if (_quoteExists) (_fee) = abi.decode(_metadata, (uint256, uint256));
+      else _fee = 3000; // TODO some default
```

Comment:
Are we taking fee here too? If not, I would just revert (if fee, then it's the whole twap thingy again?)

Reply:
this would be the twap case (yes, needed). _fee here is the param used to determine the pool being used. 
I'm cool with going with the lowest fee value, which will cater towards highly liquid pairs.